File: Lucy/2025/4-W03/1926_Nearest_Exit_from_Entrance_in_Maze.ts
function nearestExit(maze: string[][], entrance: number[]): number {
  const queue: [number, number, number][] = [];
  queue.push([entrance[0], entrance[1], 0]); // [row, col, path]

  const ROW = maze.length;
  const COL = maze[0].length;
  const EMPTY = ".";

  const direction = [
    [1, 0],
    [-1, 0],
    [0, 1],
    [0, -1],
  ];
  const visited: boolean[][] = Array.from({ length: ROW }, () => new Array(COL).fill(false));

  const getCoordinates = (row: number, col: number): string => `${row},${col}`;

  while (queue.length > 0) {
    const [row, col, currentPath] = queue.shift()!;

    if (row === 0 || row === ROW - 1 || col === 0 || col === COL - 1) {
      if (getCoordinates(row, col) !== getCoordinates(entrance[0], entrance[1])) {
        return currentPath;
      }
    }

    for (const [dr, dc] of direction) {
      const [nr, nc] = [row + dr, col + dc];

      if (nr >= 0 && nr < ROW && nc >= 0 && nc < COL && maze[nr][nc] === EMPTY && !visited[nr][nc]) {
        visited[nr][nc] = true;
        queue.push([nr, nc, currentPath + 1]);
      }
    }
  }

  return -1;
}

File: Meli/LeetCode 75/1926. Nearest Exit from Entrance in Maze.py
class Solution:
    def nearestExit(self, maze: List[List[str]], entrance: List[int]) -> int:
        from collections import deque
        rows, cols = len(maze), len(maze[0])
        queue = deque()
        queue.append((entrance[0], entrance[1], 0))  # (row, col, steps)
        
        directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]  # up, down, left, right

        while queue:
            r, c, steps = queue.popleft()
            
            # Skip if it's not an empty cell
            if maze[r][c] != '.':
                continue

            # Mark the cell as visited
            maze[r][c] = '+'

            # Check if it's an exit (on the border and not the entrance)
            if (r != entrance[0] or c != entrance[1]) and (r == 0 or c == 0 or r == rows - 1 or c == cols - 1):
                return steps

            # Explore neighbors
            for dr, dc in directions:
                nr, nc = r + dr, c + dc
                if 0 <= nr < rows and 0 <= nc < cols and maze[nr][nc] == '.':
                    queue.append((nr, nc, steps + 1))

        return -1